### PR TITLE
switch to use smart pointer in parser

### PR DESCRIPTION
**Pull request by lixupeng**
fix issue #523. The unnecessary null pointer checks mainly exist in include/parser. I replace the raw pointers with unique_ptr to avoid explicit delete operations in destructor.

### CODE REVIEW

**Comment by hzxa21** [comment]
@lixupeng Have you run the test using the new optimizer after your changes by commenting out the NEW_OPTIMIZER flag?

**Comment by apavlo** [comment]
@lixupeng Can you resolve these conflicts and run the tests with the new optimizer?

**Comment by apavlo** [comment]
@lixupeng Can you resolve these conflicts and run the tests with the new optimizer?

**Comment by apavlo** [comment]
@lixupeng Can you resolve these conflicts and run the tests with the new optimizer?

**Comment by lixupeng** [comment]
@hzxa21 Yes, it passes the tests.

File: src/planner/create_plan.cpp [hunk 1]
```diff
@@ -50,13 +50,13 @@ CreatePlan::CreatePlan(parser::CreateStatement *parse_tree) {
         
       type::TypeId val = col->GetValueType(col->type);
 
-      LOG_TRACE("Column name: %s; Is primary key: %d", col->name, col->primary);
+      LOG_TRACE("Column name: %s; Is primary key: %d", col->name.get(), col->primary);
```
Comment (apavlo) [inline src/planner/create_plan.cpp]: This won't compile under TRACE mode. Should be `col->name.get().c_str()`

File: src/planner/create_plan.cpp [hunk 2]
```diff
@@ -50,13 +50,13 @@ CreatePlan::CreatePlan(parser::CreateStatement *parse_tree) {
...
 
       // Check main constraints
       if (col->primary) {
         catalog::Constraint constraint(ConstraintType::PRIMARY, "con_primary");
         column_contraints.push_back(constraint);
-        LOG_TRACE("Added a primary key constraint on column \"%s\"", col->name);
+        LOG_TRACE("Added a primary key constraint on column \"%s\"", col->name.get());
```
Comment (apavlo) [inline src/planner/create_plan.cpp]: Change to `col->name.get().c_str()`

File: src/planner/insert_plan.cpp [hunk 3]
```diff
@@ -147,7 +147,7 @@ InsertPlan::InsertPlan(
         if (query_columns_cnt < table_columns_cnt) {
           for (size_t col_cntr = 0; col_cntr < table_columns_cnt; col_cntr++) {
             auto col = table_columns[col_cntr];
-            if (std::find(query_columns->begin(), query_columns->end(), col.GetName())
+            if (std::find_if(query_columns->begin(), query_columns->end(), [&col](std::unique_ptr<char[]>& x){return col.GetName() == x.get();})
```
Comment (apavlo) [inline src/planner/insert_plan.cpp]: This is really gross.:

`[&col](std::unique_ptr<char[]>& x){return col.GetName() == x.get();`

The casting seems wrong and overly complicated. Spread it out over multiple lines.

File: test/binder/binder_test.cpp [hunk 4]
```diff
@@ -96,11 +96,11 @@ TEST_F(BinderCorrectnessTest, SelectStatementTest) {
   // Check select_list
   LOG_INFO("Checking select list");
   auto tupleExpr =
-      (expression::TupleValueExpression*)(*selectStmt->select_list)[0];
+      (expression::TupleValueExpression*)(*selectStmt->select_list)[0].get();
```
Comment (apavlo) [inline test/binder/binder_test.cpp]: Use `dynamic_cast`

File: test/binder/binder_test.cpp [hunk 5]
```diff
@@ -151,7 +151,7 @@ TEST_F(BinderCorrectnessTest, SelectStatementTest) {
   binder.reset(new binder::BindNodeVisitor(txn));
   selectSQL = "SELECT * FROM A, B as A";
   parse_tree = parser.BuildParseTree(selectSQL);
-  selectStmt = (parser::SelectStatement*)(parse_tree->GetStatements().at(0));
+  selectStmt = (parser::SelectStatement*)(parse_tree->GetStatements().at(0).get());
```
Comment (apavlo) [inline test/binder/binder_test.cpp]: Dynamic cast.

File: test/binder/binder_test.cpp [hunk 6]
```diff
@@ -126,11 +126,11 @@ TEST_F(BinderCorrectnessTest, SelectStatementTest) {
   // Check Group By and Having
   LOG_INFO("Checking group by");
   tupleExpr =
-      (expression::TupleValueExpression*)selectStmt->group_by->columns->at(0);
+      (expression::TupleValueExpression*)selectStmt->group_by->columns->at(0).get();
   EXPECT_EQ(tupleExpr->GetBoundOid(),
             make_tuple(db_oid, tableA_oid, 0));  // A.a1
   tupleExpr =
-      (expression::TupleValueExpression*)selectStmt->group_by->columns->at(1);
+      (expression::TupleValueExpression*)selectStmt->group_by->columns->at(1).get();
```
Comment (apavlo) [inline test/binder/binder_test.cpp]: Dynamic cast.

File: test/binder/binder_test.cpp [hunk 7]
```diff
@@ -166,7 +166,7 @@ TEST_F(BinderCorrectnessTest, SelectStatementTest) {
   binder.reset(new binder::BindNodeVisitor(txn));
   selectSQL = "SELECT * FROM A, A as AA where A.a1 = AA.a2";
   parse_tree = parser.BuildParseTree(selectSQL);
-  selectStmt = (parser::SelectStatement*)(parse_tree->GetStatements().at(0));
+  selectStmt = (parser::SelectStatement*)(parse_tree->GetStatements().at(0).get());
```
Comment (apavlo) [inline test/binder/binder_test.cpp]: More dynamic cast. Same for all the other ones too.

File: test/executor/insert_test.cpp [hunk 8]
```diff
@@ -72,67 +72,67 @@ TEST_F(InsertTests, InsertRecord) {
   strcpy(name, "TEST_TABLE");
   auto table_ref = new parser::TableRef(TableReferenceType::NAME);
   parser::TableInfo *table_info = new parser::TableInfo();
-  table_info->table_name = name;
+  table_info->table_name.reset(name);
```
Comment (apavlo) [inline test/executor/insert_test.cpp]: This seems wrong.

File: test/executor/insert_test.cpp [hunk 9]
```diff
@@ -72,67 +72,67 @@ TEST_F(InsertTests, InsertRecord) {
   strcpy(name, "TEST_TABLE");
```
Comment (apavlo) [inline test/executor/insert_test.cpp]: Why strcpy?

File: src/include/optimizer/operators.h [hunk 10]
```diff
@@ -147,13 +147,13 @@ class LogicalGroupBy : public OperatorNode<LogicalGroupBy> {
 class LogicalInsert : public OperatorNode<LogicalInsert> {
  public:
   static Operator make(
-      storage::DataTable *target_table, std::vector<char *> *columns,
-      std::vector<std::vector<peloton::expression::AbstractExpression *> *> *
+      storage::DataTable *target_table, std::vector<std::unique_ptr<char[]>> *columns,
```
Comment (apavlo) [inline src/include/optimizer/operators.h]: @hzxa21 @chenboy @luochenUmich Why do we have to use raw char arrays? Can we use `std::string` instead?

File: src/include/optimizer/operators.h [hunk 11]
```diff
@@ -147,13 +147,13 @@ class LogicalGroupBy : public OperatorNode<LogicalGroupBy> {
...
+      std::vector<std::unique_ptr<std::vector<std::unique_ptr<peloton::expression::AbstractExpression>>>> *
```
Comment (apavlo) [inline src/include/optimizer/operators.h]: You are already in the `peloton` namespace, so this seems unnecessary. Change to `expression::AbstractExpression`

File: src/include/optimizer/operators.h [hunk 12]
```diff
@@ -179,10 +179,10 @@ class LogicalDelete : public OperatorNode<LogicalDelete> {
 class LogicalUpdate : public OperatorNode<LogicalUpdate> {
  public:
   static Operator make(storage::DataTable *target_table,
-                       std::vector<peloton::parser::UpdateClause*> updates);
+                       std::vector<std::unique_ptr<peloton::parser::UpdateClause>>* updates);
```
Comment (apavlo) [inline src/include/optimizer/operators.h]: You changed this from a "vector of raw pointers" to a "pointer to a vector of unique pointers". Why? Can we just pass the vector by reference?

File: src/include/parser/analyze_statement.h [hunk 13]
```diff
@@ -65,10 +56,11 @@ class AnalyzeStatement : public SQLStatement {
     v->Visit(this);
   }
 
-  parser::TableRef* analyze_table;
-  std::vector<char*>* analyze_columns;
+  std::unique_ptr<parser::TableRef> analyze_table;
+  std::unique_ptr<std::vector<std::unique_ptr<char[]>>> analyze_columns;
 
   const std::string INVALID_NAME = "";
```
Comment (apavlo) [inline src/include/parser/analyze_statement.h]: @AllisonWang What is this used for?

File: src/include/parser/insert_statement.h [hunk 14]
```diff
@@ -76,13 +45,13 @@ class InsertStatement : SQLStatement {
   }
 
   InsertType type;
-  std::vector<char*>* columns;
-  std::vector<std::vector<peloton::expression::AbstractExpression*>*>*
+  std::unique_ptr<std::vector<std::unique_ptr<char[]>>> columns;
+  std::unique_ptr<std::vector<std::unique_ptr<std::vector<std::unique_ptr<peloton::expression::AbstractExpression>>>>>
```
Comment (apavlo) [inline src/include/parser/insert_statement.h]: Switch to `expression::AbstractExpression`

File: src/include/parser/table_ref.h [hunk 15]
```diff
@@ -31,48 +31,48 @@ class JoinDefinition;
...
-  TableInfo* table_info_ = nullptr;
+  std::unique_ptr<TableInfo> table_info_ = nullptr;
 
-  char* alias;
+  std::unique_ptr<char[]> alias;
 
   SelectStatement* select;
-  std::vector<TableRef*>* list;
-  JoinDefinition* join;
+  std::unique_ptr<std::vector<std::unique_ptr<TableRef>>> list;
```
Comment (apavlo) [inline src/include/parser/table_ref.h]: This seems unnecessary to wrap the `std::vector` in a `unique_ptr`

File: src/include/parser/table_ref.h [hunk 16]
```diff
@@ -31,48 +31,48 @@ class JoinDefinition;
...
+  std::unique_ptr<JoinDefinition> join;
 
   // Convenience accessor methods
-  inline bool HasSchema() { return schema != NULL; }
+  inline bool HasSchema() { return schema != nullptr; }
 
   // Get the name of the database of this table
   inline const char* GetDatabaseName() const {
```
Comment (apavlo) [inline src/include/parser/table_ref.h]: Returning a raw pointer defeats the purpose of using smart pointers inside of the classes.

File: src/include/parser/update_statement.h [hunk 17]
```diff
@@ -60,37 +53,20 @@ class UpdateStatement : public SQLStatement {
...
   virtual void Accept(SqlNodeVisitor* v) const override {
     v->Visit(this);
   }
 
   // TODO: switch to char* instead of TableRef
-  TableRef* table;
-  std::vector<UpdateClause*>* updates;
-  expression::AbstractExpression* where = nullptr;
+  std::unique_ptr<TableRef> table;
+  std::unique_ptr<std::vector<std::unique_ptr<UpdateClause>>> updates;
```
Comment (apavlo) [inline src/include/parser/update_statement.h]: Same comment here. Not sure why the vector needs to be in a pointer.

File: src/include/optimizer/operators.h (hunk 10 above)
Comment (hzxa21) [reply src/include/optimizer/operators.h]: We can use std::string. The constructor of InsertPlan uses vector<char*>*, so we use char* also in the logical operator. To make the codes be more C++, we should also avoid char* in the plan node.

**Comment by apavlo** [comment]
@hzxa21 I agree with your points.

@lixupeng Can you fix this up quickly today and then we can merge it after Patrick approves?

**Comment by lixupeng** [comment]
@apavlo @hzxa21 Hi, I think it is ok now. I replace char* by std::string and use std::vector directly now. Could you have a look again?

**Comment by lixupeng** [comment]
@apavlo This PR is ready for review again!

**Comment by seojungmin** [comment]
Looks good to me. Just two peloton code style issues: format with 80 column, and put the & and * operator adjacent to a variable name.  Actually, some of these problems were there originally, so this PR itself can be merged even without fixing them.

**Comment by apavlo** [comment]
@lixupeng Fix the conflicts and then send an email to the mailing list to notify people that it is ready for review.

@seojungmin Thanks!

**Pull request merged**